File: 104/maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func nextDepthNode(nodeSlice []*TreeNode) (nextNodeSlice []*TreeNode) {
	for _, node := range nodeSlice {
		if nil != node.Left {
			nextNodeSlice = append(nextNodeSlice, node.Left)
		}
		if nil != node.Right {
			nextNodeSlice = append(nextNodeSlice, node.Right)
		}
	}
	return nextNodeSlice
}

func maxDepth(root *TreeNode) int {
	var maxDepth int
	if nil == root {
		return 0
	}

	nextNodeSlice := []*TreeNode{root}
	for len(nextNodeSlice) > 0 {
		maxDepth++
		nextNodeSlice = nextDepthNode(nextNodeSlice)
	}
	return maxDepth
}

File: 104/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def __init__(self):
        self.max_depth = -1

    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.dfs(root, 0)
        return self.max_depth

    def dfs(self, root, depth):
        if None == root:
            self.max_depth = max(self.max_depth, depth)
            return
        self.dfs(root.left, depth+1)
        self.dfs(root.right, depth+1)
